### PR TITLE
[WIP] Audio Queue Manager

### PR DESCRIPTION
**Pull request by davidmroth**
Added an audio queue manager to fix Audio Player playback issue (were multiple audio stream played simultaneously). 

Known issues:
- [x] Playback error (multiple audio streams playing at once)
- [x] Audio didn't stop when keyword was detected when the "News Briefing" was playing
- [x] Stop, Pause, Play, Resume audio player not working
- [ ] Media playing causes false positive of keyword detection

### CODE REVIEW

**Comment by renekliment** [comment]
Thanks for the update.
Please:
- **DON'T** close / open new pull requests - you can just add commits to the existing and we can track changes; this way we don't see the changes (code and comments); when all issues are resolved, we can always squash the PR to be one commit; 
- `queue_audio` and `play_audio` should be also in `audio.py`, but renamed without the `_audio`
- do you need `import Queue` in `main.py`? 
- don't change indentation on/around lines you don't edit

Thanks!

**Comment by davidmroth** [comment]
Thanks for the feedback!  Already working on the suggested changes.

**Comment by davidmroth** [comment]
FYI, the two open items are existing issues. I want to make sure my PR isn't held up as a result.

- Stop, Pause, Play, Resume audio player not working
- Audio player causes false positive of keyword detection (due to audio playing while active listening is on).
- 

I'll work on these items after the PR is deployed.

**Comment by renekliment** [comment]
@davidmroth I see. Feel free to file an issue in the tracker for those if you want to.

**Comment by davidmroth** [comment]
modified: main.py
        * Renamed audio player functions
        * Moving audio player subsystem and global components to separate files
    
 modified file: alexapi/player.py
        * Dedicated Audio Player (separate from Alexa voice response player) with queue handling
        * Implemented audio queueing
          
 new file: alexapi/player_state.py
        * Moved shared audio state variable to a dedicated file so they can be used globally
              - Configuration
              - Debugging
              - GPIO

 deleted: alexapi/audio.py
        * Renamed to alexapi/player.py

File: src/alexapi/player.py
Comment (renekliment) [inline src/alexapi/player.py]: This line and down (I don't know how to select multiple lines):
Get rid of these encapsulations. Just create objects in `main.py` that call the stuff directly.

File: src/alexapi/player.py
Comment (renekliment) [inline src/alexapi/player.py]: If you need to pass callbacks, do it via constructors ^, don't do this global stuff.

File: src/alexapi/player_state.py
Comment (renekliment) [inline src/alexapi/player_state.py]: This whole thing should be in `player.py`. The state of the player is a part of it.

File: src/alexapi/player.py
Comment (renekliment) [inline src/alexapi/player.py]: Do the object creation in `main.py`, where you can pass stuff via constructors.

File: src/alexapi/player.py
Comment (renekliment) [inline src/alexapi/player.py]: This does nothing.

File: src/alexapi/player.py
Comment (renekliment) [inline src/alexapi/player.py]: Call this _vlc_, _vlcInstance_ or something. Just _instance_ sounds like it's a static class variable that holds the instance in case of a singleton design pattern.

File: src/alexapi/player.py
Comment (renekliment) [inline src/alexapi/player.py]: Same as above.

File: src/alexapi/shared.py
Comment (renekliment) [inline src/alexapi/shared.py]: Get rid of this file. 
- `bcolors` should be in a separate file - [like this](https://github.com/renekliment/AlexaPiNG/blob/816f46d5901637da8ed1689b755b4f133b69aeda/src/alexapi/bcolors.py) - it is unfortunately not merged yet, but will, be, you can use that single file
- from `main.py`, pass the `config` via constructors - this is gonna be for example used when reading audio output device from config
- don't use global `debug` variable in the module - use `self.config['debug']`in the classes or something like that
- in `main.py` before passing `config` in the module (like [here](https://github.com/renekliment/AlexaPiNG/blob/816f46d5901637da8ed1689b755b4f133b69aeda/src/main.py), it's unfortunately not merged yet): 
```
if 'debug' not in config:
	config['debug'] = debug
``` 
- when you've got `debug` like this you don't need the command line parsing stuff
- put `import RPi.GPIO as GPIO` in `player.py` until #30 gets merged ... it's silly, but that's all we can do for now
- whoa, file gone

File: src/main.py
Comment (renekliment) [inline src/main.py]: `pstate.currVolume += int(vol_token)`

**Review by renekliment — CHANGES_REQUESTED** [review]
So this is quite the update!
I have a few requests, I sincerely hope they don't discourage you from contributing :) I'll be thrilled to merge this PR once it's resolved.

A few of them are our fault, because #30 isn't merged yet, but I hope you can use a few things from that PR that I've pointed out.

Apart from code comments, I have one question. From a very quick look it looks like there are two players - what's the reason for that? I thought that there's gonna be a single queue, but in `main.py` you call two distinct functions instead of `play_audio`.

Thank you!

**Comment by renekliment** [comment]
So #36 finally got merged. Are you okay solving the merge conflict? If not, let me know and I'll do that after you solve the proposed changes.

Also, this means that your PR has to pass the tests (it's errors-only pylint test only so far) in order to get merged.

**Comment by renekliment** [comment]
And #30 got merged, which might produce further conflicts. Do you need any help with that? What's your status on this? I'm planning a huge code style fix PR, which would make anything hard to merge, so I'd like to merge this before that. Thanks.

**Comment by davidmroth** [comment]
Sorry, was busy with some other things. I'll get to work on this on tomorrow and shoot for having this done by Friday.

Thank you for your feedback and patience!

**Comment by renekliment** [comment]
Awesome to hear that. Thank you!

**Comment by renekliment** [comment]
@davidmroth 
Hey! I'm very sorry, but we have a lot of audio issues and I've decided not to wait and I have opened #68 to support playback_handlers and while I was at it I've implemented a queue there.

I've seen in your PR you have some additional features if I'm correct. Feel free to add those once #68 gets merged. Design opinions are also welcome, since I'm well aware that the current implementation is not ideal.

I've seen that you're working on the new API, which is awesome! This might also take off a bit from your shoulders so you can focus on the API if you want to. The thing I'd like to suggest is you stay up-to-date with our `master`, so you don't need to waste time doing things your way only to find out you have to rewrite it later, because it's already been resolved in `master`.

I'm sorry again and hope it's fine. I have to admit, I took a little inspiration from your PR, so it did help and I hope we'll see more contributions from you (such as the API), it's greatly appreciated!

**Comment by renekliment** [comment]
P.S.: If you wanna _developer-talk_, join us at Gitter - link in the README.

**Pull request closed without merging**
